AC-484 changing aria- to data-, a to button

clrux opened the pull request:
# [AC-484](https://openedx.atlassian.net/browse/AC-484)

This work completes work with the LMS calculator and changes `aria-activedescendent` to `data-calculator-hint` so as not to conflict with any ARIA values. Also, using ARIA here was incorrect use, but we do need to store a value here to be referenced in the script and tests.

## Sandbox

https://clrux-ac-484.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/courseware/d8a6192ade314473a78242dfeedfbf5b/edx_introduction/ Calculator should already be enabled.

## Reviewers

- [x] @cptvitamin 
- [x] @cahrens or TNL

cptvitamin commented:
@clrux I think the attribute gets added here:
https://github.com/edx/edx-platform/blob/5db315bebfaf50fe17927e2445170bb20e697a21/lms/static/coffee/src/calculator.coffee

cptvitamin commented:
@clrux yeah, lets get a sandbox on this

cahrens commented on lms/static/coffee/fixtures/calculator.html:
```diff
@@ -1,13 +1,13 @@
 <ul>
   <li class="calc-main">
-    <a href="#" class="calc" title="Open Calculator" role="button" aria-controls="calculator_wrapper" aria-expanded="false">Open Calculator</a>
+    <button class="calc" title="Open Calculator" aria-controls="calculator_wrapper" aria-expanded="false">Open Calculator</button>
```
Translate the strings in this file.

clrux replied:
👍

clrux commented:
@cahrens and @cptvitamin this is ready for your review at your earliest convenience.

cahrens commented:
Uck, this calculator makes me unhappy in so many ways. For example, when I press "Enter" to execute the equation I have written, it instead brings up a "cheat sheet" of information, the top of which is cut off. How are you supposed to use the calculator with the keyboard only? How are you supposed to view the cheat sheet without zooming out in the browser?

clrux commented:
I hear ya. If you can find me approval to remove it entirely... ;)

cahrens commented:
@nasthagiri and Shelby (github handle?), we'd love to kill off the calculator. I'd love to know how many people actually use it-- I don't suppose we have any analytics data capturing that?

cahrens commented:
@sstack22 any chance we can kill off the calculator? :)

cahrens commented on lms/static/coffee/fixtures/calculator.html:
```diff
@@ -1,13 +1,13 @@
 <ul>
   <li class="calc-main">
-    <a href="#" class="calc" title="Open Calculator" role="button" aria-controls="calculator_wrapper" aria-expanded="false">Open Calculator</a>
+    <button class="calc" title="${_('Open Calculator')" aria-controls="calculator_wrapper" aria-expanded="false">${_('Open Calculator')</button>
```
Why does the button need a title given that it has text?

clrux replied:
It doesn't. I think the text is off-screen, which is probably why I added the title.

Going to hold off here until I get a yay or nay for removing this altogether...

cahrens replied:
We shouldn't need both though. If the text is offscreen, then can't we just have the title?

cahrens commented on lms/static/coffee/fixtures/calculator.html:
```diff
@@ -16,13 +16,13 @@
-                <li class="hint-item" id="hint-integers" tabindex="-1"><p><span class="bold">Integers:</span> 2520</p></li>
-                <li class="hint-item" id="hint-decimals" tabindex="-1"><p><span class="bold">Decimals:</span> 3.14 or .98</p></li>
+                <li class="hint-item" id="hint-integers" tabindex="-1"><p><span class="bold">${_('Integers:')</span> 2520</p></li>
+                <li class="hint-item" id="hint-decimals" tabindex="-1"><p><span class="bold">${_('Decimals:')</span> 3.14 or .98</p></li>
```
Translate "3.14 or .98" I guess.

clrux replied:
Can you translate numbers?

cahrens replied:
I was reacting to the "or"

cahrens commented on lms/static/coffee/fixtures/calculator.html:
```diff
@@ -16,13 +16,13 @@
         </div>
-        <input id="calculator_button" type="submit" title="Calculate" arial-label="Calculate" value="=" tabindex="-1" />
-        <input type="text" id="calculator_output" readonly tabindex="-1" />
+        <input id="calculator_button" type="submit" title="${_('Calculate')" aria-label="${_('Calculate')" value="=" />
```
Why does there need to be both a title and an aria-label?

clrux replied:
We don't need both.

sstack22 commented:
I don't have usage data on the calculator tool. Anyone have ideas on how we could get this? I do know at least a few courses use it.

nasthagiri commented:
Is the calculator tool an xBlock?  If not, does it have an endpoint that it accesses, which we can check in splunk?

cahrens commented:
@sstack22  I imagine we can easily so how many courses have calculator enabled. But the more interesting question is how many students are actually using it. I can look at that in splunk, as it calls a "calculate" endpoint.

cptvitamin commented:
@cahrens you have to move focus to the equal sign to compute the answer.  My guess is your focus was on the "info" or help icon when you hit Enter/Space.

I'm not going to weigh in either way on whether or not the calc should go away (its not turned on my default so anyone who enables it, must want it. We should check to see how many active courses are using it.)  WRT the fixture file, I believe that is just for testing.  These frequently don't stay i sync with the actual implementation used in production.  I think that is what is going on here.   That fixture file should emulate the code that is used in production for it to be worthwhile.  I agree with @cahrens that it is unlikely that this button needs offscreen text *and* sr text.  I think title alone should suffice.

clrux commented:
If this is something we want to keep around, maybe I can make this a small Hackathon project today/tomorrow - retool it from the ground up?

clrux commented:
One thing I'm noticing, and this might be @cahrens issue, is if I type "4+4" and press Enter, instead of providing the calculation, it shows the help menu. I have to explicitly Tab twice to the equal sign to get the calculation.

Might be just a matter of handling that Enter/Return press. I'll work with it today.

clrux commented:
@cahrens I've updated the code to submit the equation on "Enter/Return" rather than show the Hints. I also updated tabindexes so that the result receives focus so the actual result is announced properly.

Finally I tweaked the styles so that the button is clear, along with there result container.

clrux commented:
jenkins run bokchoy

clrux commented:
@cptvitamin @cahrens Ready for review.

cahrens commented:
@clrux Enter is still bringing up the cheat sheet for me on Chrome (on the sandbox). So either there is a bug, or you did not update the sandbox.

clrux commented:
SMH. Forgot to update the sandy. One sec... sorry.

clrux commented:
@cahrens updated!

clrux commented:
@cahrens @cptvitamin Ready for you guys again. Sandy is updated.

cahrens commented on lms/static/coffee/fixtures/calculator.html:
```diff
@@ -1,29 +1,30 @@
+                                </span>
+                            </p>
+                        </li>
+                        <li class="hint-item" id="hint-integers" tabindex="-1"><p><span class="bold">${_('Integers:')</span> ${_('2520')</p></li>
```
The numerical values will need translator comments so that they translators have a clue as to what to do with them.

clrux replied:
In the fixture, or the actual `toggle_calculator.html` file?

clrux replied:
The lines in the main file aren't translated (don't need to be). I should update the fixture though.

cahrens replied:
Uck, I didn't notice that this is a fixture file! It makes sense though, as I couldn't figure out how to view this text in the calculator.

Nothing in the fixture file should be translated, as it just produces unnecessary work for the translators. Please remove the _ usages in this file.

cahrens commented on lms/templates/calculator/toggle_calculator.html:
```diff
@@ -4,141 +4,150 @@
-    <span class="icon fa fa-calculator" aria-hidden="true"></span>
-    <span class="utility-control-label sr">${_("Open Calculator")}</span>
-  </button>
+    <button title="${_('Open Calculator')}" aria-controls="calculator_wrapper" aria-expanded="false" class="calc utility-control-button">
```
It would be best for you to fix the safe template issues with this file while you are editing it. Please take a look at the quality report when it finishes running.

clrux replied:
You got it. I guess we are all-in on this calculator, eh? :)

cahrens commented on lms/templates/calculator/toggle_calculator.html:
```diff
@@ -4,141 +4,150 @@
-  </div>
+                <input type="submit" id="calculator_button" value="=" aria-label="${_('Calculate')}">
+                <label for="calculator_output" class="calc-output-label sr">${_('Result')}</label>
+                <input type="text" id="calculator_output" tabindex="-1" readonly />
```
Should you be able to tab into the output area? I find the current behavior after you press Enter a bit confusing because you don't see where focus is.

clrux replied:
It's kind of weird. I'm sending focus so the results can be announced, but the `readonly` is funky. Let me talk to Mark and see what he says.

clrux commented:
@cahrens I am sending focus to the result container, but it's `readonly` so the focus ring doesn't show up. You mentioned this being confusing. I've added styles that reflect when a result has been given (green glow on result field) to indicate a successful calculation. and ease the usability/experience concern.

Anything else related to the calculator that doesn't pertain to the scope of this work will be addressed later. For now, please review its current state. The sandbox has been updated.

@cptvitamin You too, sir.

I'd love to merge this today or tomorrow.

cahrens commented:
Thanks @clrux, it looks better with the focus ring. There are 5 safecommit violations-- can you please fix them?

cahrens commented on lms/static/coffee/src/calculator.coffee:
```diff
@@ -192,16 +201,19 @@ class @Calculator
 
   handleClickOnDocument: (e) ->
     @hideHint()
+    $('#calculator_output').removeClass('has-result')
```
You don't need to call this here-- hideHint also removes the class (line 108).

cahrens commented on lms/static/coffee/src/calculator.coffee:
```diff
@@ -192,16 +201,19 @@ class @Calculator
     if @hintPopup.hasClass 'shown'
       @hideHint()
       @hintButton.attr('aria-expanded', false)
+      $('#calculator_output').removeClass('has-result')
```
Again, this is not needed due to hideHint.

cahrens commented on lms/static/coffee/src/calculator.coffee:
```diff
@@ -192,16 +201,19 @@ class @Calculator
     else
       @showHint()
       @hintButton.attr('aria-expanded', true)
+      $('#calculator_output').removeClass('has-result')
```
This is not needed here-- showHint also removes the class.

clrux commented:
@cahrens Unnecessary calls removed. Safe template fixes implemented.

cptvitamin commented on lms/templates/calculator/toggle_calculator.html:
```diff
@@ -1,144 +1,158 @@
+                <div class="help-wrapper">
+                    <p class="sr" id="hint-instructions">${_('Use the arrow keys to navigate the tips or use the tab key to return to the calculator')}</p>
+                    
+                    <button type="button" title="${_('Hints')}" id="calculator_hint" aria-haspopup="true" aria-expanded="false" aria-controls="calculator_input_help" aria-describedby="hint-instructions"></button>
```
Hey @clrux this is unrelated to your PR, but Screen reader users will encounter this text before they move focus to the button (that references it via aria-describedby).  This really only affects VO users, who will encounter everything in the DOM, and will occasionally affect Windows screen reader users when they are in read mode, but I still think we should address it.  Instead of help text on the button, why don't we make this screen reader text the first text in the help modal.  Perhaps put it on line 28 below.

cahrens commented on lms/templates/calculator/toggle_calculator.html:
```diff
@@ -1,144 +1,157 @@
+                    <button type="button" title="${_('Hints')}" id="calculator_hint" aria-haspopup="true" aria-expanded="false" aria-controls="calculator_input_help"></button>
+                    <ul id="calculator_input_help" class="help" aria-hidden="true">
+                        <li class="hint-item" id="hint-moreinfo">
+                            <p class="sr" id="hint-instructions">${_('Use the arrow keys to navigate the tips or use the tab key to return to the calculator')}</p>
```
Is this message accurate? If I bring up the hints with the keyboard, using the arrow keys doesn't navigate the tips-- it scrolls the window behind the hints. And if press Tab, I'm not sure where my focus goes (hints do not close).

clrux replied:
I moved this "hint" inside of the "hint panel" because it was previously being announced prematurely. There is a lot of work we _could_ do to the calculator, but my aim is to fix the one tiny thing. This has blown up beyond the initial scope.

Tomorrow I will get anything out of order, in proper order and update the sandbox for a final review. Note the initial scope of work was to simply change an invalid ARIA attribute. In my quest to "fix all the things!" I've let this get way out of control.

In short, I'll verify tomorrow that things work correctly, then have you guys review the safe template updates (understandably necessary) and the ARIA change. Everything else should work as it has before, funky or not.

cahrens replied:
Sounds good. FWIW, I did review the safe template commit that's already on this PR, and it looked good.

cahrens commented:
Have you notice that the close button is not focusable? I don't see any way to close the calculator with the keyboard.

cahrens commented on lms/templates/calculator/toggle_calculator.html:
```diff
@@ -1,144 +1,157 @@
+        <form id="calculator">
+            <div class="input-wrapper">
+                <label for="calculator_input" class="label-calc-input sr">${_('Enter equation')}</label>
+                <input type="text" id="calculator_input" title="${_('Calculator Input Field')}" aria-describedby="calculator_output" />
```
Is it correct to link the input field to the output field with the aria-described by?

aria-describedby="calculator_output"

I would imagine it could be confusing, as all the numbers would just run together.

cahrens commented:
Sorry, just a few more questions (see above).

clrux commented:
@cahrens @cptvitamin I've gone through and done a little bit more cleanup. The `aria-describedby` was removed as its's no longer necessary since we're sending focus to the output/result field. I've removed the unused `id` on the hint text.

Keystroke programming didn't change. You can close the hints by toggling the same button you used to show them (not ideal). You can close the calculator by Shift-Tab'ing back up the DOM to the same toggle button you used to open it.

The original scope of this work was to correct an invalid ARIA attribute and make the templates safe. We got overzealous and went above and beyond the call of duty and made other improvements. But in short, if you could just scope this final review to the ARIA attribute and safe templating for thumbs, I'd appreciate it.

cahrens commented:
👍 (squash commits as usual)

cptvitamin commented:
I'm comfortable giving 👍  on the original scope of work but we should think about getting rid of this or fixing it up proper at some point.  At least now its not creating any automated a11y test failures.  Thanks for taking care of that.

clrux commented:
Thanks. If the calculator is something we're going to keep, I'd like to rebuild it from the ground up. It might be something we can separate from LMS, maybe create a plugin or something. Anyways, thanks for the thumbs.

clrux commented:
jenkins run bokchoy
